gobblin-core/src/main/java/gobblin/instrumented/Instrumented.java

```diff
@@ -131,6 +138,85 @@ public static MetricContext getMetricContext(State state, Class<?> klazz, List<T
   }
 
   /**
+   * Copies a metric context and returns a new context with different name, and possibly different tags.
+   * @param context Context to copy.
+   * @param tags New tags to apply to context. Repeated keys will override old tags.
+   * @param name Name of the new context. If absent or empty, will modify old name by adding a random integer at the end.
+   * @return Copied context.
+   */
+  public static MetricContext copyMetricContext(MetricContext context, List<Tag<?>> tags, Optional<String> name) {
+
+    String newName = name.orNull();
+
+    if (Strings.isNullOrEmpty(newName)) {
+      int randomId = RAND.nextInt();
+
+      String oldName = context.getName();
+      List<String> splitName = Splitter.on(".").splitToList(oldName);
+      if (StringUtils.isNumeric(Iterables.getLast(splitName))) {
+        splitName.set(splitName.size() - 1, Integer.toString(randomId));
+      } else {
+        splitName.add(Integer.toString(randomId));
+      }
+      newName = StringUtils.join(splitName, ".");
+    }
+
+    MetricContext.Builder builder = context.getParent().isPresent() ?
+        context.getParent().get().childBuilder(newName) :
+        MetricContext.builder(newName);
+    return builder.addTags(context.getTags()).addTags(tags).build();
+  }
+
+  /**
+   * Determines whether an object or, if it is a decorator, any object on its lineage, is already instrumented.
+   * @param obj Object to analyze.
+   * @return Whether the object is instrumented.
+   */
+  public static boolean isLineageInstrumented(Object obj) {
+    List<Object> lineage = Lists.newArrayList(obj);
+    if(obj instanceof Decorator) {
+      lineage = DecoratorUtils.getDecoratorLineage(obj);
```

Comment:
Seems `DecoratorUtils.getDecoratorLineage` already handles both cases. So you don't need to do the `if`.